File: rental/src/main/java/io/axoniq/demo/bikerental/rental/ui/RentalController.java
package io.axoniq.demo.bikerental.rental.ui;

import io.axoniq.demo.bikerental.coreapi.rental.BikeStatus;
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.queryhandling.QueryGateway;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ThreadLocalRandom;

@RestController
@RequestMapping("/")
public class RentalController {

    public static final String FIND_ALL_QUERY = "findAll";
    public static final String FIND_ONE_QUERY = "findOne";
    private static final List<String> LOCATIONS = Arrays.asList("Amsterdam", "Paris", "Vilnius", "Barcelona", "London", "New York", "Toronto", "Berlin", "Milan", "Rome", "Belgrade");
    private final CommandGateway commandGateway;
    private final QueryGateway queryGateway;

    public RentalController(CommandGateway commandGateway, QueryGateway queryGateway) {
        this.commandGateway = commandGateway;
        this.queryGateway = queryGateway;
    }

    @PostMapping
    public CompletableFuture<Void> generateBikes(@RequestParam("bikes") int bikeCount,
                                                 @RequestParam(value = "bikeType") String bikeType) {
        CompletableFuture<Void> all = CompletableFuture.completedFuture(null);
        for (int i = 0; i < bikeCount; i++) {
            all = CompletableFuture.allOf(all,
                    commandGateway.send(new RegisterBikeCommand(UUID.randomUUID().toString(), bikeType, randomLocation())));
        }
        return all;
    }

    @PostMapping("/requestBike")
    public CompletableFuture<String> requestBike(@RequestParam("bikeId") String bikeId, @RequestParam("renter") String renter) {
        // TODO Implement
        return null;
    }

    @PostMapping("/returnBike")
    public CompletableFuture<String> returnBike(@RequestParam("bikeId") String bikeId, @RequestParam("location") String location) {
        // TODO Implement
        return null;
    }

    @GetMapping("/bikes")
    public CompletableFuture<List<BikeStatus>> findAll() {
        // TODO Implement
        return null;
    }

    @GetMapping("/bikes/{bikeId}")
    public CompletableFuture<BikeStatus> findStatus(@PathVariable("bikeId") String bikeId) {
        // TODO Implement
        return null;
    }

    private String randomLocation() {
        return LOCATIONS.get(ThreadLocalRandom.current().nextInt(LOCATIONS.size()));
    }

}

File: rental/src/main/kotlin/io/axoniq/demo/bikerental/rental/ui/RentalController.kt
package io.axoniq.demo.bikerental.rental.ui

import io.axoniq.demo.bikerental.coreapi.rental.BikeStatus
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand
import io.axoniq.demo.bikerental.coreapi.rental.RequestBikeCommand
import io.axoniq.demo.bikerental.coreapi.rental.ReturnBikeCommand
import io.axoniq.demo.bikerental.rental.query.FindAllQuery
import io.axoniq.demo.bikerental.rental.query.FindOneQuery
import org.axonframework.commandhandling.gateway.CommandGateway
import org.axonframework.extensions.kotlin.query
import org.axonframework.extensions.kotlin.queryMany
import org.axonframework.queryhandling.QueryGateway
import org.springframework.web.bind.annotation.*
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ThreadLocalRandom


@RestController
@RequestMapping("/")
class RentalController(
    private val commandGateway: CommandGateway,
    private val queryGateway: QueryGateway
) {

    @PostMapping
    fun generateBikes(
        @RequestParam("bikes") bikeCount: Int,
        @RequestParam("bikeType") bikeType: String
    ): CompletableFuture<Void> = CompletableFuture.allOf(
        *(0 until bikeCount)
            .map {
                commandGateway.send<Any>(
                    RegisterBikeCommand(bikeId = UUID.randomUUID().toString(), bikeType, randomLocation())
                )
            }
            .toTypedArray()
    )

    @PostMapping("/requestBike")
    fun requestBike(
        @RequestParam("bikeId") bikeId: String,
        @RequestParam("renter") renter: String
    ): CompletableFuture<String> =
        commandGateway.send(RequestBikeCommand(bikeId, renter))

    @PostMapping("/returnBike")
    fun returnBike(
        @RequestParam("bikeId") bikeId: String,
        @RequestParam("location") location: String
    ): CompletableFuture<String> =
        commandGateway.send(ReturnBikeCommand(bikeId, location))

    @GetMapping("/bikes")
    fun findAll(): CompletableFuture<List<BikeStatus>> =
        queryGateway.queryMany(FindAllQuery)

    @GetMapping("/bikes/{bikeId}")
    fun findStatus(@PathVariable("bikeId") bikeId: String): CompletableFuture<BikeStatus> =
        queryGateway.query(FindOneQuery(bikeId))

    private fun randomLocation(): String =
        LOCATIONS[ThreadLocalRandom.current().nextInt(LOCATIONS.size)]

    companion object {
        private val LOCATIONS = listOf(
            "Amsterdam",
            "Paris",
            "Vilnius",
            "Barcelona",
            "London",
            "New York",
            "Toronto",
            "Berlin",
            "Milan",
            "Rome",
            "Belgrade"
        )
    }
}
